Exclude retired pools from the `ListStakePools` API function result.

jonathanknowles opened the pull request:
# Issue Number

#1937 

# Overview

This PR:

- [x] Applies a very simple filter to the Shelley `ListStakePools` API operation that excludes all pools with a retirement epoch earlier than or equal to the current epoch.
- [x]  Adds an integration test to verify that if pool _**p**_ has already retired, it will **not** be listed by `ListStakePools`.

# Future Improvements

The current pool tracking implementation (which predates this PR) has two potential areas for improvement:

- Reducing the number of database queries that need to be executed:
    See: https://jira.iohk.io/browse/ADP-383
- Performing garbage collection of retired pools from the database:
    See: https://jira.iohk.io/browse/ADP-376

These are **not** tackled in this PR, but are instead recorded in the above tickets, so that we may schedule time to fix them at a later date.

jonathanknowles committed:
Extract out function `listStakePools_` in Shelley API server.

jonathanknowles committed:
Add `EpochNo` filter argument to Shelley `listStakePools` API server function.

jonathanknowles committed:
Pass the current epoch into the Shelley `listStakePools` API server function.

jonathanknowles committed:
Filter out retired pools from Shelley `listStakePools` API function.

jonathanknowles committed:
Add reminders to use more efficient database queries.

Add reminders to use more efficient database queries when generating the
complete list of non-retired pools.

See: https://jira.iohk.io/projects/ADP/issues/ADP-383

jonathanknowles committed:
Add new stake pool to Shelley integration test framework.

Generated by following the instructions here:

https://docs.cardano.org/projects/cardano-node/en/latest/stake-pool-operations/node_keys.html

The pool id appears to be arbitrary.

jonathanknowles committed:
Make comments compatible with Haddock.

Unfortunately, Haddock doesn't support documentation comments here.

So we make these comments into ordinary Haskell comments.

jonathanknowles committed:
Add `PoolConfig` for a pool that will retire very early.

jonathanknowles committed:
Make it possible to sort `StakePoolMetadata` values into order.

jonathanknowles committed:
Make the "contains pool metadata" test pass.

jonathanknowles committed:
Add comments to `testPoolConfigs` to explain intention.

piotr-iohk commented on lib/core-integration/src/Test/Integration/Scenario/API/Shelley/StakePools.hs:
```diff
@@ -679,16 +679,21 @@ spec = do
+                        let metadataActual = Set.fromList $
+                                mapMaybe (fmap getApiT . view #metadata) pools
+                        metadataActual
+                            `shouldSatisfy` (`Set.isSubsetOf` metadataPossible)
```
Nitpicking (:sweat_smile:), but maybe we could add a stronger assertion that the list of returned pools does not contain GPD?

jonathanknowles replied:
> Nitpicking (sweat_smile), but maybe we could add a stronger assertion that the list of returned pools does not contain GPD?

Unfortunately we can't currently do this, as there is no guarantee that pool `GPD` will be the one to be assigned with an early retirement epoch. It could be _any_ of the four pools that gets retired early.

jonathanknowles replied:
I suspect we could solve this problem entirely by changing our test framework to allow pools to be registered and retired dynamically, as in: https://jira.iohk.io/browse/ADP-385

piotr-iohk commented on lib/shelley/src/Cardano/Wallet/Shelley/Launch.hs:
```diff
@@ -1122,6 +1122,33 @@ operators = unsafePerformIO $ newMVar
           ]
       )
+    , ( PoolId $ unsafeFromHex
+          "5a7b67c7dcfa8c4c25796bea05bcdfca01590c8c7612cc537c97012bed0dec36"
```
:tada:

KtorZ replied:
Actually no :/ ... The pool id is a hash of the operator verification key. I am not sure how this successfully got merged but it's now causing issue.

piotr-iohk reviewed (commented):
There is one problem I think. Playing around with the CI cluster I've observed that the pool scheduled to be retired in epoch `n` is still visible through the whole epoch `n` and disappears from the list in epoch `n+1`. While it is visible one can still join it but gets:
```
{"code":"no_such_pool"
,"message":"I couldn't find any stake pool with the given id: b45768c1a2da4bd13ebcaa1ea51408eda31dcc21765ccbd407cda9f2"}
```

I guess pool retired to be scheduled in `n` should not be shown from `n` onwards...

jonathanknowles commented on lib/shelley/src/Cardano/Wallet/Shelley/Pools.hs:
```diff
@@ -166,17 +175,35 @@ newStakePoolLayer gp nl db@DBLayer {..} = StakePoolLayer
             $ combineDbAndLsqData (slotParams gp) lsqData dbData
+      where
+        epochIsInFuture :: EpochNo -> Bool
+        epochIsInFuture = (>= currentEpoch)
```
Surely, this should be:

```suggestion
        epochIsInFuture = (> currentEpoch)
```

jonathanknowles replied:
Fixed in 6ec42fbe22762b91d2607f1b2bdab8e2312d6eb7.

jonathanknowles committed:
Correct off-by-one error in function `epochIsInFuture`.

In response to review feedback:

https://github.com/input-output-hk/cardano-wallet/pull/1945/files#r460378458

piotr-iohk reviewed (approved):
LGTM!

Corner case that may be considered in future PR/story:
 - pool A schedule to retire in `n`
 - wallet joins A in `n-1` (shows as delegation next: `n + 1`, target: A)
 - pool disappears from the list in `n` (wallet still shows as delegation next: `n + 1`, target: A)
 - since `n + 1` wallet shows as delegation active, target: A
